### DISCUSSION TITLE
How libvips can use NVIDIA GPU on AWS g2 series VM/machines

### DISCUSSION THREAD

**Discussion by vikasjio** [libvips/libvips#2112] [Q&A]
Hi,

I am using libvips in my program for image transcoding.
I am using a VM/Server which is having a Nvidia GPU for performance enhamcement.

But surprisingly my GPU is not being used and showing 0% usage; despite normal/intel CPUs/cores on the server/VM going upto 100% usage under heavy load conditions. we are using AWS g2.2x and g2.8x VMs as our server for image transcoding.

I searched libvips source code etc but found no option/way to spread image processing load to nvidia GPU.

Request if someone can guide on how to spread image processing load to nvidia GPU.

Awaiting,
Vikas

**Comment by jcupitt** [comment]
Hi @vikasjio, libvips has no direct support for GPU processing, it's CPU-only. Some of the libraries that libvips uses can use your GPU, for example some experimental branches of openjpeg have some GPU support, but not libvips.

In any case, a GPU would be unlikely to make any speed difference. It depends on the type of processing you are doing, but if you are just resizing and changing format, processing speed will be limited by image decode and recode, and a GPU cannot help there. 

Do you have liborc enabled? If this is turned on, libvips will do run-time code generation for your server's vector units, and this can speed things up measurably.

In my experience, the first thing to do for performance tuning is to define a simple benchmark which is representative of your workload, profile it, and see where time is being spent. I find callgrind very helpful.

**Comment by vikasjio** [comment]
Thanks @jcupitt ,
I am only doing 1. resizing, 2. croping, 3. image rotation and 4. changing format (usually to jpeg) only and nothing else.
But my load is high - thousands of such image transcodings per second (on few MBs image files). And my server CPU reaches 100% in hundreds of load, so i will have to use multiple Servers. (appreciable - libvips is stable in this load)
So, GPU will not make speed/processing improvement in this case?

I think I have enabled liborc. I assume so from the below log in my config.log (in config.status section):
accelerate loops with orc:      yes
  (requires orc-0.4.11 or later)

plz correct if I mis-understood liborc enablement.

Thanks again,
Vikas

**Comment by jcupitt** [comment]
Yes, it sounds like you have liborc enabled. Can you share your code? I could have a look if you like, there might be some improvement that could be made. For example, what language are you using, how are you calling vips, what image formats, which resize operation, which colour management option. 

I made an example to show the problem with GPU acceleration. This is on my laptop:

```
$ vipsheader wtc.jpg 
wtc.jpg: 9372x9372 uchar, 3 bands, srgb, jpegload
$ time vips copy wtc.jpg x.jpg
real    0m0.872s
user    0m1.496s
sys 0m0.048s
```

So 870ms to read, decompress, copy (zero time) and recompress a 10,000 x 10,000 pixel JPEG image. 

Now try running a simple operation on the same image:

```
$ more avg.mat 
3 3 9 0 
1 1 1
1 1 1
1 1 1
$ time vips conv wtc.jpg x.jpg avg.mat 
real    0m1.073s
user    0m2.468s
sys 0m0.076s
```

Now it's 1070ms. So even if GPU processing were instant, which it is not, using a GPU could only give a maximum speedup of 200ms, or about 20%. Most time is being spent in image read and write, where the GPU cannot help.

**Comment by vikasjio** [comment]
I am using C language and using no explicit color management option.

Yes, my simplified code flow is like this (mainly capturing vips interface calls):-

// read image - step1
im = vips_image_new_from_file (procInfo->ifn, NULL)

// resize image, if requested/required - step2 optional
vips_image_write(im, im2);
im_resize_linear(im, im2, width, height);

// crop, if requested/required - step3 optional
vips_image_write(im2, im3);
vips_crop(im2, &im3, (im2->Xsize - width)/2, (im2->Ysize - height)/2, width, height, NULL)

// rotate, if requested/required - step4 optional
vips_image_write(im3, im32)
VipsAngle angle = vips_autorot_get_angle( im_123 );
vips_rot( in, out, angle, NULL )

// save/deliver as jpeg - step5
vips_jpegsave_mime(im_123, NULL);

(I skipped application logic and vips related initialization, de-initialization/cleanup etc for the sake of brevity)

Please note that step1/read and step5/write are done for all the requests. And intermediate steps 2, 3, 4 are optional and depends on my input/request parameters - which have many permutation/combinations; however this sequence of flow is usually following, unrequired optional steps are simply skipped.

I understand that i am calling vips_image_write multiple times and it might be reduced, but i think this function only copies some metadata and not whole image data, so I assume it to be a light function and hence did not clubbed to keep my code simple/logically organised; however, if you think can improve performance significant, I am open to modification here as well.

I am using -Ofast option of the gcc compiler. (And I am not short on RAM side).

complete status summary section of config.log is like this:-

configure:28450: result:
- build options
  native win32:               no
  native OS X:                no
  open files in binary mode:      no
  enable debug:               no
  build deprecated components:        yes
  build docs with gtkdoc:         no
  gobject introspection:          no
  build vips7 Python binding:     yes
  install vips8 Python overrides:     no
  (requires pygobject-3.12.0 or later)
- optional dependencies
  use fftw3 for FFT:          yes
  Magick package:             MagickWand
  file import with libMagick:         yes
  accelerate loops with orc:      yes
  (requires orc-0.4.11 or later)
  ICC profile support with lcms:      yes (lcms2)
  file import with OpenEXR:       yes
  file import with OpenSlide:     yes
  (requires openslide-3.3.0 or later)
  file import with matio:         yes
  file import with cfitsio:       yes
  file import/export with libwebp:    yes
  text rendering with pangoft2:       yes
  file import/export with libpng:     yes (pkg-config libpng >= 1.2.9)
  (requires libpng-1.2.9 or later)
  file import/export with libtiff:    yes (pkg-config libtiff-4)
  file import/export with libjpeg:    yes
  image pyramid export:           yes
  (requires libgsf-1 1.14.27 or later)
  use libexif to load/save JPEG metadata: yes

Thanks,
Vikas

**Comment by jcupitt** [comment]
Could you post a complete example I can try benchmarking? There are some pieces missing from your sample code, so I can't really test it. For example, how are you making the image that you write to with `im_resize_linear`? 

You should use `vips_resize()` --- `im_resize_linear()` does not work well, and it's very slow. You shouldn't need the `vips_image_write()` calls either. You should probably be using sequential access mode.

**Comment by vikasjio** [comment]
Thanks @jcupitt ,

I will share with you complete example.
I will make it complete (even after removing intermingled non-vips stuff). I would also try with vips_resize before disturbing you further.

regards,
Vikas

**Comment by vikasjio** [comment]
Thanks @jcupitt ,

So, here is the complete compilable executable program that best reflects my usage of vips library mainly for resizing and reformatting (to jpeg), but also for croping and rotation etc.

I am compiling this program with below command:

gcc -Ofast -Wall perftxm.c  `pkg-config vips --cflags --libs`

Below is the complete perftxm.c:

```C
#define _GNU_SOURCE
#include <stdio.h>

#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <string.h>
#include <strings.h>

#include <vips/vips.h>

#define TEJ_TRUE    1
#define TEJ_FALSE   0

#define TEJ_NULL    '\0'
#define TEJ_RC_SUCCESS  0

#define TEJDL_DEBUG 1

#define tejLog  printf

#define AWS_S3_SUPPORT  1
#define HPP_MODE    1

/*
- Below is the changeable info, as its mostly input to the aplication
  */

int g_isWebRule = TEJ_TRUE;
int g_width = 200;
int g_height = 300;

// Config Info Over.

char    ifn[250];
int image_len = 0;

#define ORIENTATION_IFD0    ("exif-ifd0-Orientation")
#define ORIENTATION_IFD1    ("exif-ifd1-Orientation")

/**
 *
- Look at the exif tags and rotate the image to make it upright. The
- orientation tag is removed from @out to prevent accidental double rotation.
  *
- Returns: 0 on success, -1 Otherwise.
  _/
  int txm_autorotate(VipsImage *in, VipsImage *_out)
  {
  VipsAngle angle = vips_autorot_get_angle( in );
  
  if( angle != VIPS_ANGLE_D0 )
  {
      if( vips_rot( in, out, angle, NULL ) )
      {
          return -1;
      }
      (void) vips_image_remove( (_out), ORIENTATION_IFD0 );
      (void) vips_image_remove( (_out), ORIENTATION_IFD1 );
  
  return 0;
 }
  
  return -1;
  }

/*
- This function trascodes the info/image
  */
  int txInfo(void)
  {
  int width;
  int height;
  int intWidth;
  int intHeight;
  int isim3 = TEJ_FALSE;
  int val = 0;
  int result = -1;
  VipsImage *im;
  VipsImage *im2 = vips_image_new();
  VipsImage *im22 = vips_image_new();
  VipsImage *im3 = vips_image_new();
  VipsImage *im32 = vips_image_new();
  
  /*
    Read an image.
  */
  if (!(im = vips_image_new_from_file (ifn, NULL)))
  {
  #ifdef TEJDL_DEBUG
      tejLog("unable to open ifn %s.\n", ifn);
  #endif
      g_object_unref (im3);
      g_object_unref (im2);
      return 4041;
  }
  
  /*
    Transcode the images into a .
  */
  if(TEJ_TRUE ) // assume it true here
  {
      if(vips_image_write(im, im2))
      {
  #ifdef TEJDL_DEBUG
          tejLog("unable to image_write.\n");
  #endif
          g_object_unref (im);
          g_object_unref (im2);
          return 4041;
      }
  
  width = g_width;
  
  if( TEJ_FALSE == g_isWebRule )
  {
      // nh = nw * h / w
      height = width * im->Ysize / im->Xsize;
 
  
  #ifdef TEJDL_DEBUG
          tejLog("\norig size - width: %d, height: %d\n", im->Xsize, im->Ysize);
          tejLog("\nresize - width: %d, height: %d\n", width, height);
  #endif
  
    if( im_resize_linear(im, im2, width, height) )
      {
  
  #ifdef TEJDL_DEBUG
              tejLog("unable to resize.\n");
  #endif
              g_object_unref (im);
              g_object_unref (im2);
              return 4041;
          }
      }
      else
      {
          height = g_height;
  #ifdef TEJDL_DEBUG
          tejLog("\norig size - width: %d, height: %d\n", im->Xsize, im->Ysize);
          tejLog("\nresize - width: %d, height: %d\n", width, height);
  #endif
  
      if( (width < im->Xsize) || (height < im->Ysize) )
      {
          if( im->Xsize < im->Ysize )
          {
              intWidth = width;
              intHeight = width * im->Ysize / im->Xsize;
          }
          else
          {
              intWidth = height * im->Xsize / im->Ysize;
              intHeight = height;
          }
  
  #ifdef TEJDL_DEBUG
              tejLog("\nresize - intwidth: %d, intheight: %d\n", intWidth, intHeight);
  #endif
  
          if( im_resize_linear(im, im2, intWidth, intHeight) )
          {
  
  #ifdef TEJDL_DEBUG
                  tejLog("unable to resize.\n");
  #endif
                  g_object_unref (im);
                  g_object_unref (im2);
                  return 4041;
              }
  
          if( intWidth != intHeight )
          {
              if(vips_image_write(im2, im3))
              {
  
  #ifdef TEJDL_DEBUG
                      tejLog("unable to image_write3.\n");
  #endif
  
                  g_object_unref (im);
                  g_object_unref (im2);
                  g_object_unref (im3);
                  return 4041;
              }
  
              if(vips_crop(im2, &im3, (im2->Xsize - width)/2, (im2->Ysize - height)/2, width, height, NULL))
              {
  
  #ifdef TEJDL_DEBUG
                      tejLog("unable to crop.\n");
  #endif
                      g_object_unref (im);
                      g_object_unref (im2);
                      g_object_unref (im3);
                      return 4041;
                  }
                  isim3 = TEJ_TRUE;
              }
          }
      }
  }
  
  /*
    Write the image then destroy it.
  */
  #if defined AWS_S3_SUPPORT && defined HPP_MODE
  printf("Accept-Ranges: none\r\n");
  printf("Content-Disposition: attachment\r\n");
  #endif
  if(TEJ_TRUE ) // changed for perf test
  {
      if( TEJ_TRUE == isim3 )
      {
          if( TEJ_TRUE == g_isWebRule )
          {
              if( vips_image_write(im3, im32) )
              {
  #ifdef TEJDL_DEBUG
                  tejLog("unable to image_write32.\n");
  #endif
              }
              else
              {
                  result = txm_autorotate(im3, &im32);
              }
          }

#ifdef TEJDL_DEBUG
            tejLog("\nresult: %d; resized - width: %d, height: %d\n", result, im3->Xsize, im3->Ysize);
#endif
#if defined AWS_S3_SUPPORT && defined HPP_MODE
            val = vips_jpegsave_mime( ( result ? im3 : im32 ), NULL);  // it also prints Content-Type and Content-Length headers.
            image_len = VIPS_IMAGE_SIZEOF_IMAGE( ( result ? im3 : im32 ) );
#else
            val = vips_image_write_to_file( ( result ? im3 : im32 ) , g_ofn, NULL);
#endif
        }
        else
        {
            if( TEJ_TRUE == g_isWebRule )
            {
                if( vips_image_write(im2, im22) )
                {
#ifdef TEJDL_DEBUG
                    tejLog("unable to image_write22.\n");
#endif
                }
                else
                {
                    result = txm_autorotate(im2, &im22);
                }
            }

#ifdef TEJDL_DEBUG
            tejLog("\nresult: %d; resized - width: %d, height: %d\n", result, im2->Xsize, im2->Ysize);
#endif
#if defined AWS_S3_SUPPORT && defined HPP_MODE
            val = vips_jpegsave_mime( ( result ? im2 : im22 ), NULL);  // it also prints Content-Type and Content-Length headers.
            image_len = VIPS_IMAGE_SIZEOF_IMAGE( ( result ? im2 : im22 ) );
#else
            val = vips_image_write_to_file( ( result ? im2 : im22 ) , g_ofn, NULL);
#endif
        }
    }
    if (val)
    {
#ifdef TEJDL_DEBUG
        tejLog("unable to write.\n");
#endif

    g_object_unref (im);
    g_object_unref (im2);
    g_object_unref (im22);
    g_object_unref (im3);
    g_object_unref (im32);
    return 4041;
}

g_object_unref (im);
g_object_unref (im2);
g_object_unref (im22);
g_object_unref (im3);
g_object_unref (im32);

return TEJ_RC_SUCCESS;

}

int main(int argc, char **argv)
{

if( argc != 2 )
{
    printf("\nUsage: commandName inputImageFilePathName\n");
    return 0;
}
// setConfigAndCustInfo();
strncpy(ifn, argv[1], 250);
printf("\nifn: %s\n", ifn);

if( TEJ_RC_SUCCESS == txInfo() ) // Actual Image transcode
{
    printf("\nImage Transcoded OK!!\n");
}
else
{
    printf("\nImage Transcoding Failed!!\n");
}

return 0;

}
```

I could not find a way to attach the file, so pasted whole src code here.

Mostly varying/configurable parameters are:-

```
int g_isWebRule = TEJ_FALSE; // Or TEJ_TRUE
int g_width = 200;   // Any value from 50 to 4080; but usually 100 to 640
int g_height = 300; // Any value from 50 to 4080; but usually 100 to 640. It may or may not be in proportion or ratio of ht/wd of original image
```

Please Note that I am getting my input image Over Network; And I take it directly into file pointer mounted on my shared memory (in my real application, which I changed here for sake of simplicity and portability), so I assume minimum read time.
I also have to send the transcoded image over network (I don't store it with me), so I use directly vips_jpegsave_mime(), to avoid any time on image write.

Sorry for some delay. Let me know if you you need any more info.
Kindly advise me for any performance improvement, if possible, especially if any significant.

Regards,
Vikas

**Comment by vikasjio** [comment]
The above pasted src code seems improperly formatted.
If you want the file uploaded/emailed, kindly suggest me way/emailId.

Vikas
+91-9871054768

**Comment by vikasjio** [comment]
#define _GNU_SOURCE
#include <stdio.h>

#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <string.h>
#include <strings.h>

#include <vips/vips.h>

#define TEJ_TRUE    1
#define TEJ_FALSE   0

#define TEJ_NULL    '\0'
#define TEJ_RC_SUCCESS  0

#define TEJDL_DEBUG 1

#define tejLog  printf

#define AWS_S3_SUPPORT  1
#define HPP_MODE    1

/*
- Below is the changeable info, as its mostly input to the aplication
  */

int g_isWebRule = TEJ_TRUE;
int g_width = 200;
int g_height = 300;

// Config Info Over.

char    ifn[250];
int image_len = 0;

#define ORIENTATION_IFD0    ("exif-ifd0-Orientation")
#define ORIENTATION_IFD1    ("exif-ifd1-Orientation")

/**
 *
- Look at the exif tags and rotate the image to make it upright. The
- orientation tag is removed from @out to prevent accidental double rotation.
  *
- Returns: 0 on success, -1 Otherwise.
  _/
  int txm_autorotate(VipsImage *in, VipsImage *_out)
  {
  VipsAngle angle = vips_autorot_get_angle( in );
  
  if( angle != VIPS_ANGLE_D0 )
  {
      if( vips_rot( in, out, angle, NULL ) )
      {
          return -1;
      }
      (void) vips_image_remove( (_out), ORIENTATION_IFD0 );
      (void) vips_image_remove( (_out), ORIENTATION_IFD1 );
  
  ```
  return 0;
  ```
  
  }
  
  return -1;
  }

/*
- This function trascodes the info/image
  */
  int txInfo(void)
  {
  int width;
  int height;
  int intWidth;
  int intHeight;
  int isim3 = TEJ_FALSE;
  int val = 0;
  int result = -1;
  VipsImage *im;
  VipsImage *im2 = vips_image_new();
  VipsImage *im22 = vips_image_new();
  VipsImage *im3 = vips_image_new();
  VipsImage *im32 = vips_image_new();
  
  /*
    Read an image.
  */
  if (!(im = vips_image_new_from_file (ifn, NULL)))
  {
  #ifdef TEJDL_DEBUG
      tejLog("unable to open ifn %s.\n", ifn);
  #endif
      g_object_unref (im3);
      g_object_unref (im2);
      return 4041;
  }
  
  /*
    Transcode the images into a .
  */
  if(TEJ_TRUE ) // assume it true here
  {
      if(vips_image_write(im, im2))
      {
  #ifdef TEJDL_DEBUG
          tejLog("unable to image_write.\n");
  #endif
          g_object_unref (im);
          g_object_unref (im2);
          return 4041;
      }
  
  ```
  width = g_width;
  
  if( TEJ_FALSE == g_isWebRule )
  {
      // nh = nw * h / w
      height = width * im->Ysize / im->Xsize;
  ```
  
  #ifdef TEJDL_DEBUG
          tejLog("\norig size - width: %d, height: %d\n", im->Xsize, im->Ysize);
          tejLog("\nresize - width: %d, height: %d\n", width, height);
  #endif
  
  ```
      if( im_resize_linear(im, im2, width, height) )
      {
  ```
  
  #ifdef TEJDL_DEBUG
              tejLog("unable to resize.\n");
  #endif
              g_object_unref (im);
              g_object_unref (im2);
              return 4041;
          }
      }
      else
      {
          height = g_height;
  #ifdef TEJDL_DEBUG
          tejLog("\norig size - width: %d, height: %d\n", im->Xsize, im->Ysize);
          tejLog("\nresize - width: %d, height: %d\n", width, height);
  #endif
  
  ```
      if( (width < im->Xsize) || (height < im->Ysize) )
      {
          if( im->Xsize < im->Ysize )
          {
              intWidth = width;
              intHeight = width * im->Ysize / im->Xsize;
          }
          else
          {
              intWidth = height * im->Xsize / im->Ysize;
              intHeight = height;
          }
  ```
  
  #ifdef TEJDL_DEBUG
              tejLog("\nresize - intwidth: %d, intheight: %d\n", intWidth, intHeight);
  #endif
  
  ```
          if( im_resize_linear(im, im2, intWidth, intHeight) )
          {
  ```
  
  #ifdef TEJDL_DEBUG
                  tejLog("unable to resize.\n");
  #endif
                  g_object_unref (im);
                  g_object_unref (im2);
                  return 4041;
              }
  
  ```
          if( intWidth != intHeight )
          {
              if(vips_image_write(im2, im3))
              {
  ```
  
  #ifdef TEJDL_DEBUG
                      tejLog("unable to image_write3.\n");
  #endif
  
  ```
                  g_object_unref (im);
                  g_object_unref (im2);
                  g_object_unref (im3);
                  return 4041;
              }
  
              if(vips_crop(im2, &im3, (im2->Xsize - width)/2, (im2->Ysize - height)/2, width, height, NULL))
              {
  ```
  
  #ifdef TEJDL_DEBUG
                      tejLog("unable to crop.\n");
  #endif
                      g_object_unref (im);
                      g_object_unref (im2);
                      g_object_unref (im3);
                      return 4041;
                  }
                  isim3 = TEJ_TRUE;
              }
          }
      }
  }
  
  /*
    Write the image then destroy it.
  */
  #if defined AWS_S3_SUPPORT && defined HPP_MODE
  printf("Accept-Ranges: none\r\n");
  printf("Content-Disposition: attachment\r\n");
  #endif
  if(TEJ_TRUE ) // changed for perf test
  {
      if( TEJ_TRUE == isim3 )
      {
          if( TEJ_TRUE == g_isWebRule )
          {
              if( vips_image_write(im3, im32) )
              {
  #ifdef TEJDL_DEBUG
                  tejLog("unable to image_write32.\n");
  #endif
              }
              else
              {
                  result = txm_autorotate(im3, &im32);
              }
          }

#ifdef TEJDL_DEBUG
            tejLog("\nresult: %d; resized - width: %d, height: %d\n", result, im3->Xsize, im3->Ysize);
#endif
#if defined AWS_S3_SUPPORT && defined HPP_MODE
            val = vips_jpegsave_mime( ( result ? im3 : im32 ), NULL);  // it also prints Content-Type and Content-Length headers.
            image_len = VIPS_IMAGE_SIZEOF_IMAGE( ( result ? im3 : im32 ) );
#else
            val = vips_image_write_to_file( ( result ? im3 : im32 ) , g_ofn, NULL);
#endif
        }
        else
        {
            if( TEJ_TRUE == g_isWebRule )
            {
                if( vips_image_write(im2, im22) )
                {
#ifdef TEJDL_DEBUG
                    tejLog("unable to image_write22.\n");
#endif
                }
                else
                {
                    result = txm_autorotate(im2, &im22);
                }
            }

#ifdef TEJDL_DEBUG
            tejLog("\nresult: %d; resized - width: %d, height: %d\n", result, im2->Xsize, im2->Ysize);
#endif
#if defined AWS_S3_SUPPORT && defined HPP_MODE
            val = vips_jpegsave_mime( ( result ? im2 : im22 ), NULL);  // it also prints Content-Type and Content-Length headers.
            image_len = VIPS_IMAGE_SIZEOF_IMAGE( ( result ? im2 : im22 ) );
#else
            val = vips_image_write_to_file( ( result ? im2 : im22 ) , g_ofn, NULL);
#endif
        }
    }
    if (val)
    {
#ifdef TEJDL_DEBUG
        tejLog("unable to write.\n");
#endif

```
    g_object_unref (im);
    g_object_unref (im2);
    g_object_unref (im22);
    g_object_unref (im3);
    g_object_unref (im32);
    return 4041;
}

g_object_unref (im);
g_object_unref (im2);
g_object_unref (im22);
g_object_unref (im3);
g_object_unref (im32);

return TEJ_RC_SUCCESS;
```

}

int main(int argc, char **argv)
{

```
if( argc != 2 )
{
    printf("\nUsage: commandName inputImageFilePathName\n");
    return 0;
}
// setConfigAndCustInfo();
strncpy(ifn, argv[1], 250);
printf("\nifn: %s\n", ifn);

if( TEJ_RC_SUCCESS == txInfo() ) // Actual Image transcode
{
    printf("\nImage Transcoded OK!!\n");
}
else
{
    printf("\nImage Transcoding Failed!!\n");
}

return 0;
```

}

**Comment by jcupitt** [comment]
The simplest way is to make a gist:

https://gist.github.com/

Then paste the link here.

**Comment by vikasjio** [comment]
Thanks @jcupitt 

the gist link is below:-

https://gist.github.com/vikasjio/f9d4c778fe56174db1fe25964aaef4e8

Hope I have done it right.

Vikas

**Comment by jcupitt** [comment]
Thanks @vikasjio, I'll experiment here. 

What kinds of images are you running this on? Are they JPG or TIF or PNG? What is a typical image size (in pixels)? Are they all RGB?

**Comment by vikasjio** [comment]
Thanks @jcupitt .

Typically majority of images will be captured/shooted by modern smartphone cameras, expecting sizes in the range of 2 to 8 MBs per image. Other properties same as that typical smartphone camera generates now a days and in near future.
However, theoretically and to some minimum frequency, there can be almost any image input available on consumer electronic devices; so we have to support many input formats mainly including Jpeg/jpg, webp, png, tiff, gif, bmp, hdr are the formats that we are supporting currently, more may be added going forward.

Also, Considering your suggestions and our lateral understanding, I am already considering to move to AWS's gpu-less C4 series VMs/servers from currently used gpu series g2.

regards,
Vikas

**Comment by jcupitt** [comment]
I've made you a new load/shrink/crop/rotate program:

https://gist.github.com/jcupitt/ff6067e5d90e80394181b2357199d933

I've added some comments to explain how it works. The big changes are:
- use jpeg shrink-on-load
- load from a memory buffer rather than the filesystem
- use `vips_resize()`, not `im_resize_linear()`
- fix the intermediate images
- improve the object unref handling

With an iphone6s camera image and your code I saw:

```
$ time ./a.out ~/Desktop/IMG_1867.JPG > x.jpg
real    0m0.293s
user    0m0.308s
sys 0m0.220s
```

With this new code, I see:

```
$ time ./a.out ~/Desktop/IMG_1867.JPG > x.jpg
real    0m0.129s
user    0m0.116s
sys 0m0.012s
```

So perhaps 3x faster, since startup and shutdown will take some fixed amount. 

Quality is much better too, here's the before and after:

![x](https://cloud.githubusercontent.com/assets/580843/18207589/acf901ae-7123-11e6-852f-6507891f23f4.png)

**Comment by asilvas** [comment]
Great maintainer -- now that's help!

**Comment by strarsis** [comment]
So I could use the GPU with sharp/libvips on WSL 2?

**Comment by jcupitt** [comment]
No, libvips has no code for using the GPU, it's a pure CPU library. 

That may change in the future, of course.
